push to stable

vanhauser-thc opened the pull request:
getting the new pass manager out there and fixing unicorn issues

wtdcode committed:
Fix sample script for Unicornafl2 (#1329)

* Fix the sample script for Unicornafl2

* Make it compatible with unicornafl1

* Sync to the latest unicornafl

* Update UNICORNAFL_VERSION

* Sync to latest unicornafl to resolve linkage problem

* Switch to ucafl2

Commit:
Changes to simplify coverage code

Commit:
Added instrumentation for CMOV instructions

Commit:
Changes to default CMOV instrumentation to off

Commit:
Added test for running python

Commit:
Reserved used address space on dlclose to prevent re-use

vanhauser-thc committed:
Merge pull request #1332 from WorksButNotTested/module

Module

Commit:
Add global caching of block translation to instrumentation

vanhauser-thc committed:
Revert "remove new llvm pass manager :("

This reverts commit 55ed2a443c5c61baba37415d4087164454d8a2a8.

vanhauser-thc committed:
todo list

Teemperor committed:
LLVM 14 fixes

vanhauser-thc committed:
Merge pull request #1334 from Teemperor/llvm14

Fix compilation against LLVM 14

vanhauser-thc committed:
Merge branch 'newpm2' into dev

vanhauser-thc committed:
Merge pull request #1335 from AFLplusplus/dev

Dev

devnexen committed:
Fix build for the LLVM pass for LLVM >= 13

Due to change of inheritance for more modern LLVM versions, the overridable
members are not the same.

adamdoupe committed:
Clarify in docs that AFL_TARGET_ENV will apply to QEMU in QEMU mode, and note that QEMU_SET_ENV should be used instead in QEMU mode. Closes #1328

vanhauser-thc committed:
Merge pull request #1339 from adamdoupe/dev

Clarify in docs that AFL_TARGET_ENV will apply to QEMU in QEMU mode.

vanhauser-thc committed:
Merge pull request #1337 from devnexen/llvm_13_plus_fix

Fix build for the LLVM pass for LLVM >= 13

vanhauser-thc committed:
remove file

vanhauser-thc committed:
Merge pull request #1340 from AFLplusplus/dev

Dev

vanhauser-thc committed:
todo list for llvm

martinclauss committed:
Update fuzzing_in_depth.md

add libdesock as an additional way to ease the fuzzing for network servers (#1341)

vanhauser-thc committed:
Merge pull request #1342 from martinclauss/patch-1

Update fuzzing_in_depth.md

vanhauser-thc committed:
Revert "Merge branch 'newpm2' into dev"

This reverts commit fee1acf7e6096533f1aa8cd74035bed21c90fdf6, reversing
changes made to 675d17d737ee5dee88766d9c181567771592c94c.

vanhauser-thc committed:
allow for -c0 also for src targets

vanhauser-thc committed:
code format

vanhauser-thc committed:
code format

vanhauser-thc committed:
ported cmplog-instructions-pass

vanhauser-thc committed:
port cmplog-switches-pass

vanhauser-thc committed:
ported cmplog-routines-pass

vanhauser-thc committed:
port instrumentation/afl-llvm-dict2file

schumilo committed:
fix broken code format

schumilo committed:
update submodules

schumilo committed:
write nyx crash logs to AFL++ work directory

schumilo committed:
fix compile error on non linux targets

vanhauser-thc committed:
Merge pull request #1344 from schumilo/dev

support Nyx crash logs

vanhauser-thc committed:
code format

vanhauser-thc committed:
sancov will be only for llvm11+

mchesser committed:
Fix issue with trim in shared memory mode

mchesser committed:
Fix off by one bounds check

`is_hex` reads two bytes but caller previously only ensured that one byte was in bounds.

mchesser committed:
Fix buffer overrun in `rtn_extended_encoding`

`idx + i` can range from `0` to `buf.len`, but the memcpy may try and write to offsets from `idx + i` to `idx + 2 * i`.

vanhauser-thc committed:
new compiler performance options

vanhauser-thc committed:
nits

vanhauser-thc committed:
Merge pull request #1346 from mchesser/patches

Fix Shm trim bug and OOB accesses during Redqueen mutations.

vanhauser-thc committed:
Merge pull request #1347 from WorksButNotTested/cache

Add global caching of block translation to instrumentation

vanhauser-thc committed:
test

cd80 committed:
fix typo for error message when using LD_PRELOAD

vanhauser-thc committed:
Merge pull request #1352 from cd80/patch-1

fix typo in error message when using LD_PRELOAD

vanhauser-thc committed:
make pcguard work

vanhauser-thc committed:
make LTO pass work too plus some fixes

vanhauser-thc committed:
fixes for llvm < 11

vanhauser-thc committed:
Merge branch 'dev' into newpm2

vanhauser-thc committed:
Merge pull request #1353 from AFLplusplus/newpm2

new pass manager

vanhauser-thc committed:
merge fixes

vanhauser-thc committed:
forgot one pass

vanhauser-thc committed:
fix split pass loading

vanhauser-thc committed:
our own PCGUARD mode not for llvm 10.0.1

vanhauser-thc committed:
fix LTO with cmplog and laf

rwmjones committed:
docs: Document linker scripts change required in some projects

AFL instrumentation modifies the binary to export extra symbols which
must be visible if the main binary dlopen()s an instrumented DLL.
This change was required by nbdkit since it uses a linker script to
limit the symbols exported from the main binary:

https://gitlab.com/nbdkit/nbdkit/-/commit/355cacb812126f5e37a2468f6939821b15017373

Signed-off-by: Richard W.M. Jones <rjones@redhat.com>

vanhauser-thc committed:
Merge pull request #1355 from rwmjones/document-linker-scripts

docs: Document linker scripts change required in some projects

fuzzah committed:
fix early return in AFL_TARGET_ENV

vanhauser-thc committed:
Merge pull request #1359 from fuzzah/fix-1333-afl-target-env

fix early return in AFL_TARGET_ENV

vanhauser-thc committed:
update qemu mode

vanhauser-thc committed:
cleanup

clchrlls committed:
fix installation packages for debian 10

clchrlls committed:
missing column

clchrlls committed:
Not sure the fix is needed only for debian 10

clchrlls committed:
reverted to simply changing gcc version retrieval

clchrlls committed:
put back the title of the page deleted by mistake

vanhauser-thc committed:
Merge pull request #1363 from clchrlls/stable

fix installation packages commands for debian 10

domenukk committed:
fix unicornafl test

vanhauser-thc committed:
fix qemu sigunaction tests

vanhauser-thc committed:
fix instrumentlist for LTO with clang < 13

vanhauser-thc committed:
remove debug

hexcoder- committed:
avoid compiler warning (format-truncation), needed for make DEBUG=1 all

hexcoder- committed:
ignore unused variable to let make DEBUG=1 distrib succeed

domenukk committed:
Some scan-build fixes

domenukk commented on docs/Changelog.md:
```diff
@@ -12,6 +12,9 @@ sending a mail to <afl-users+subscribe@googlegroups.com>.
   - fix */build_...sh scripts to work outside of git
   - new custom_mutator: libafl with token fuzzing :)
   - afl-fuzz:
+    - when you just want to compile once and set CMPLOG, then just
```
I feel like that could be default, and the binary (and other instrumentations) could tell afl-fuzz it supports inline cmplog on handshake?
Different binaries to get more perf is an advanced use-case

vanhauser-thc replied:
hmm that could be an interesting option indeed. and how to tell afl-fuzz not to enable cmplog then (if the target has cmplog compiled in)?

domenukk replied:
F0r example `AFL_AUTO_CMPLOG=0` or `-c -` or something similar?

vanhauser-thc replied:
yeah maybe ... lets put that on the TODO list

domenukk commented on docs/fuzzing_in_depth.md:
```diff
@@ -817,9 +835,9 @@ Here are some of the most important caveats for AFL++:
+  simple code changes to make them behave in a more traditional way. Preeny or libdesock may
   offer a relatively simple option, too - see:
-  [https://github.com/zardus/preeny](https://github.com/zardus/preeny)
+  [https://github.com/zardus/preeny](https://github.com/zardus/preeny) or [https://github.com/fkie-cad/libdesock](https://github.com/fkie-cad/libdesock)
```
We could port over FitM QEMU ideas/network emu at some point in time or hint at it here (https://github.com/fgsect/FitM)

vanhauser-thc replied:
that would be cool and could be then also applied to libafl right?

domenukk replied:
LibAFL would be a perfect fit for this, yes, but it needs time and someone to do it

vanhauser-thc replied:
I mean the patch to qemu could be used for both (or maybe the fitm qemu changes are already generic and dont need any changes?)

domenukk replied:
For LibAFl it'd be cleaner to reimplement the FitM changes on top of the LibAFL's syscall API imho

domenukk commented on instrumentation/SanitizerCoveragePCGUARD.so.cc:
```diff
@@ -992,6 +1011,15 @@ bool ModuleSanitizerCoverage::InjectCoverage(Function &             F,
                   FixedVectorType::get(Int32PtrTy, elements);
               Value *x, *y;
 
+              if (!FunctionGuardArray) {
```
@vanhauser-thc `scan-build` complained that `FunctionGuardArray` may be set to `null` here, maybe either fixing the root cause or exiting the build here could be a better solution (and at two other locations)

vanhauser-thc replied:
these are false positives in scan-build. it is impossible that FunctionGuardArray is NULL, otherwise the function would not be called ... I think better than adding that if (!...) rather put a scan-build comment that this is a false positive? (sure there is a way to flag these)

domenukk replied:
I prefer the if, the flow looks sane I think so there is a chance this might happen, and if not, it doesn't hurt

devnexen committed:
frida mode fix arm64 build

rwmjones committed:
GNUmakefile.llvm: Avoid creating circular links

Building into a DESTDIR gives circular links such as:

lrwxrwxrwx. 1 rjones rjones 17 Mar 31 18:09 /home/rjones/rpmbuild/BUILDROOT/american-fuzzy-lop-4.00c-3.git285a5cb3.fc37.x86_64/usr/lib64/afl/afl-compiler-rt.o -> afl-compiler-rt.o

I have to admit I've no idea what the original link command is trying
to achieve, but it's clearly wrong and deleting it fixes the problem.

Signed-off-by: Richard W.M. Jones <rjones@redhat.com>

vanhauser-thc committed:
Merge pull request #1372 from devnexen/frida_arm64_build_fix

frida mode fix arm64 build

vanhauser-thc committed:
Merge pull request #1373 from rwmjones/no-circular-links

GNUmakefile.llvm: Avoid creating circular links
